### ISSUE TITLE
Pipeline with Redis Cluster

### ISSUE THREAD

**Issue by jangitje** [sewenew/redis-plus-plus#350]
I want to use pipeline with Redis Cluster. This only seems to work in case I use a hashtag (like {hashtagh} in key. 
My question is if this is still required in latest release of Redis.

**Comment by sewenew** [comment]
Yes, because all keys in a pipeline must be in the same slot. So in order to ensure that commands in a pipeline are sent to the right node, you need to use hashtag to specify the slot / node.

Regards

**Comment by jangitje** [comment]
I do not want to appear a wise guy but is it absolutely sure this is still the case?I have a colleague just using Hiredis reassuring me this would no longer be required.

**Comment by sewenew** [comment]
AFAIK, yes, that's still the case.

Regards

**Comment by mrmgxxxx** [comment]
key-1 --> slot 1
key-1 --> slot 1
key-1 --> slot 1

**Comment by jangitje** [comment]
I am pretty sure pipeline support is there for hiredis-c;luster and that's probably how they do it. Wonder though if that doesn't make it slower.

**Comment by mrmgxxxx** [comment]
> I am pretty sure pipeline support is there for hiredis-c;luster and that's probably how they do it. Wonder though if that doesn't make it slower.

All right, i add some debug log in main logic of pipeline，1000 keys to hmget，and it exec the pipeline 1000 times

**Comment by mrmgxxxx** [comment]
there is a way i think to make it, get pipeline object with pool in redis cluster node level，e.g. 3 nodes cluster ，it only exec pipeline 3 times to get all keys

**Comment by mrmgxxxx** [comment]
I did the complete test, it works, but i modified the source code

make RedisCluster _pool public(or add getxxx method), and fetch pool by user level, and build the Pipeline object,

finally, 1000 keys only exec 3 times(there is a  3 nodes cluster), and the performance metric data much better now

**Comment by sewenew** [comment]
> I am pretty sure pipeline support is there for hiredis-c;luster and that's probably how they do it. Wonder though if that doesn't make it slower.

@jangitje Do you mean [Nordix/hiredis-cluster](https://github.com/Nordix/hiredis-cluster)? That's not the official c client. AFAIK, the official one is [hiredis](https://github.com/redis/hiredis) does not have cluster support.

I do not check the source code of hiredis-cluster, but I think it might split those keys internally, and send them to different node.

Regards

**Comment by sewenew** [comment]
> there is a way i think to make it, get pipeline object with pool in redis cluster node level，e.g. 3 nodes cluster ，it only exec pipeline 3 times to get all keys

@mrmgxxxx For each node, there's a connection pool. However, that's implementation details, so you'd better not use it in application code.

If you has already known the key distribution, i.e. which keys belongs to which node, you can create a `Redis` object to the specific node, and send all command in a single pipeline.

Also if performance is a big concern, you can try the redis-plus-plus' [async interface](https://github.com/sewenew/redis-plus-plus#async-interface). With async interface, you don't even need Pipeline. Instead, you can send all commands with `AsyncRedisCluster`, and for each command you'll get a `future` object. After sending all commands, you wait all related `future`s, that works as the pipeline.

```
vector<future<vector<OptionalString>>> res;
for (auto &cmd : commands) {
    res.push_back(async_redis_cluster.hmget(cmd.key, cmd.field));  // send commands without waiting for result
}
vector<OptionalString> real_res;
for (auto &fut : res) {   // waiting all result
    real_res.push_back(fut.get());
}
```

Regards

**Comment by sewenew** [comment]
By the way, do you create pipeline with a new connection or [with connection in pool](https://github.com/sewenew/redis-plus-plus#create-pipeline-without-creating-new-connection)? If you create new connection for each pipeline, you'll get very bad performance.

Regards

**Comment by mrmgxxxx** [comment]
> > there is a way i think to make it, get pipeline object with pool in redis cluster node level，e.g. 3 nodes cluster ，it only exec pipeline 3 times to get all keys
> 
> @mrmgxxxx For each node, there's a connection pool. However, that's implementation details, so you'd better not use it in application code.
> 
> If you has already known the key distribution, i.e. which keys belongs to which node, you can create a `Redis` object to the specific node, and send all command in a single pipeline.
> 
> Also if performance is a big concern, you can try the redis-plus-plus' [async interface](https://github.com/sewenew/redis-plus-plus#async-interface). With async interface, you don't even need Pipeline. Instead, you can send all commands with `AsyncRedisCluster`, and for each command you'll get a `future` object. After sending all commands, you wait all related `future`s, that works as the pipeline.
> 
> ```
> vector<future<vector<OptionalString>>> res;
> for (auto &cmd : commands) {
>     res.push_back(async_redis_cluster.hmget(cmd.key, cmd.field));  // send commands without waiting for result
> }
> vector<OptionalString> real_res;
> for (auto &fut : res) {   // waiting all result
>     real_res.push_back(fut.get());
> }
> ```
> 
> Regards

i create the pipeline object base on the connection pool in reuse mode，i know what you mean and i do not want to break the internal details in hack mode either，i will check the async interface，which maybe a better way to choose i hope

**Comment by jangitje** [comment]
> 

My requirements is a streaming application with at max of around 50.000 lookups per second.
In my original setup - no Redis involved - the was a problem, I had throttling and occasionally crashing.
I swapped to Redis Cluster without pipeline and performance was better but still not good enough.
After going to pipeline performance is actually excellent, far within requirements. And actually I do create new connections. I guess I got a little scared because of the warnings. But I will certainly try :-)
The hashtag for pipeline with Redis Cluster is part of looking for the ultimate configuration.

Thanks,
Jan
